### ISSUE TITLE
Testing PySpark solid - AttributeError: 'NoneType' object has no attribute 'setCallSite'

### ISSUE THREAD

**Issue by tseader** [dagster-io/dagster#1980]
I'm receiving an error when attempting to test a Solid that uses Pyspark - `AttributeError: 'NoneType' object has no attribute 'setCallSite'`.  I'm not sure how to resolve this issue as it has only ever occurred with Dagster and the [StackOverflow guidance](https://stackoverflow.com/questions/50606059/attributeerror-nonetype-object-has-no-attribute-setcallsite), nor the related [SPARK-27335](https://issues.apache.org/jira/browse/SPARK-27335?jql=text%20~%20%22setcallsite%22) suggested solutions have worked.  Note that the issue exists when attempting to `collect()` on the Spark DataFrame, though the issue is not limited only to collect().

Also note that I tried using @spark_solid with the same results, though that attempt is not reflected in this issue.  

I have attempted to provide the exact setup I am using to aid in resolution. Below are the setup steps, code being tested, and testing output from pytest.

# Full VENV setup notes:
```shell
python3 -m venv venv
source venv/bin/activate
python --version # Python 3.6.8
pip install --upgrade pip
pip install dagster dagit dagster-pyspark pyspark pytest
pip list
Package             Version    
------------------- -----------
aniso8601           7.0.0      
argh                0.26.2     
attrs               19.3.0     
bleach              3.1.0      
certifi             2019.11.28 
chardet             3.0.4      
Click               7.0        
coloredlogs         10.0       
contextlib2         0.6.0.post1
dagit               0.6.6      
dagster             0.6.6      
dagster-graphql     0.6.6      
dagster-pyspark     0.6.6      
dagster-spark       0.6.6      
decorator           4.4.1      
defusedxml          0.6.0      
entrypoints         0.3        
enum-compat         0.0.3      
Flask               1.1.1      
Flask-Cors          3.0.8      
Flask-GraphQL       2.0.1      
Flask-Sockets       0.2.1      
funcsigs            1.0.2      
future              0.18.2     
gevent              1.4.0      
gevent-websocket    0.10.1     
graphene            2.1.8      
graphql-core        2.2.1      
graphql-relay       2.0.1      
graphql-server-core 1.1.3      
graphql-ws          0.3.0      
graphviz            0.13.2     
greenlet            0.4.15     
humanfriendly       4.18       
idna                2.8        
importlib-metadata  1.3.0      
ipython-genutils    0.2.0      
itsdangerous        1.1.0      
Jinja2              2.10.3     
jsonschema          3.2.0      
jupyter-core        4.6.1      
MarkupSafe          1.1.1      
mistune             0.8.4      
more-itertools      8.0.2      
nbconvert           5.6.1      
nbformat            4.4.0      
packaging           19.2       
pandocfilters       1.4.2      
pathtools           0.1.2      
pip                 19.3.1     
pluggy              0.13.1     
promise             2.2.1      
py                  1.8.0      
py4j                0.10.7     
Pygments            2.5.2      
pyparsing           2.4.5      
pyrsistent          0.15.6     
pyspark             2.4.4      
pytest              5.3.1      
PyYAML              3.13       
requests            2.22.0     
Rx                  1.6.1      
setuptools          40.6.2     
six                 1.13.0     
SQLAlchemy          1.3.11     
testpath            0.4.4      
toposort            1.5        
traitlets           4.3.3      
urllib3             1.25.7     
watchdog            0.9.0      
wcwidth             0.1.7      
webencodings        0.5.1      
Werkzeug            0.16.0     
zipp                0.6.0
```

# Python Test Code
```python
from pyspark.sql import SparkSession
from pyspark.sql import DataFrame as SparkDataFrame
from dagster_pyspark import pyspark_resource
from dagster import execute_solid, ModeDefinition, solid


@solid
def simple_solid(context):
    spark: SparkSession = context.resources.pyspark.spark_session
    data = [("1", "2")]
    df = spark.createDataFrame(data, ["col1", "col2"])
    return df


def test_simple_pyspark_decorator():
    res = execute_solid(simple_solid, mode_def=ModeDefinition(
        resource_defs={'pyspark': pyspark_resource}))
    df: SparkDataFrame = res.output_value()
    assert isinstance(df, SparkDataFrame)
    print(df.collect())
```

# Test execution and output

```text
$ python -m pytest tests/int/test_dagster_spark.py -s -vvvv
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.6.8, pytest-5.3.1, py-1.8.0, pluggy-0.13.1 -- /home/z092789/code/treasury-ml/venv/bin/python
cachedir: .pytest_cache
rootdir: /home/z092789/code/treasury-ml, inifile: pytest.ini
plugins: cov-2.8.1, pylint-0.14.1, mockito-0.0.4
collected 1 item                                                                                                                                                                                          

tests/int/test_dagster_spark.py::test_simple_pyspark_decorator 19/12/11 09:20:12 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
2019-12-11 09:20:13 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7932e7ba-97fe-45e2-b9c2-a1e20f687ec7 - PIPELINE_START - Started execution of pipeline "ephemeral_simple_solid_solid_pipeline".
2019-12-11 09:20:13 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7932e7ba-97fe-45e2-b9c2-a1e20f687ec7 - ENGINE_EVENT - Executing steps in process (pid: 12721)
 event_specific_data = {"metadata_entries": [["pid", null, ["12721"]], ["step_keys", null, ["{'simple_solid.compute'}"]]]}
2019-12-11 09:20:13 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7932e7ba-97fe-45e2-b9c2-a1e20f687ec7 - STEP_START - Started execution of step "simple_solid.compute".
               solid = "simple_solid"
    solid_definition = "simple_solid"
            step_key = "simple_solid.compute"
2019-12-11 09:20:15 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7932e7ba-97fe-45e2-b9c2-a1e20f687ec7 - STEP_OUTPUT - Yielded output "result" of type "Any". (Type check passed).
 event_specific_data = {"intermediate_materialization": null, "step_output_handle": ["simple_solid.compute", "result"], "type_check_data": [true, "result", null, []]}
               solid = "simple_solid"
    solid_definition = "simple_solid"
            step_key = "simple_solid.compute"
2019-12-11 09:20:15 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7932e7ba-97fe-45e2-b9c2-a1e20f687ec7 - STEP_SUCCESS - Finished execution of step "simple_solid.compute" in 1.88s.
 event_specific_data = {"duration_ms": 1876.2479840079322}
               solid = "simple_solid"
    solid_definition = "simple_solid"
            step_key = "simple_solid.compute"
2019-12-11 09:20:15 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7932e7ba-97fe-45e2-b9c2-a1e20f687ec7 - ENGINE_EVENT - Finished steps in process (pid: 12721) in 1.88s
 event_specific_data = {"metadata_entries": [["pid", null, ["12721"]], ["step_keys", null, ["{'simple_solid.compute'}"]]]}
2019-12-11 09:20:15 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7932e7ba-97fe-45e2-b9c2-a1e20f687ec7 - PIPELINE_SUCCESS - Finished execution of pipeline "ephemeral_simple_solid_solid_pipeline".
FAILED

================================================================================================ FAILURES =================================================================================================
______________________________________________________________________________________ test_simple_pyspark_decorator ______________________________________________________________________________________

    def test_simple_pyspark_decorator():
        res = execute_solid(simple_solid, mode_def=ModeDefinition(
            resource_defs={'pyspark': pyspark_resource}))
        df: SparkDataFrame = res.output_value()
        assert isinstance(df, SparkDataFrame)
>       print(df.collect())

tests/int/test_dagster_spark.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib64/python3.6/site-packages/pyspark/sql/dataframe.py:533: in collect
    with SCCallSiteSync(self._sc) as css:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pyspark.traceback_utils.SCCallSiteSync object at 0x7fc7e9547550>

    def __enter__(self):
        if SCCallSiteSync._spark_stack_depth == 0:
>           self._context._jsc.setCallSite(self._call_site)
E           AttributeError: 'NoneType' object has no attribute 'setCallSite'

venv/lib64/python3.6/site-packages/pyspark/traceback_utils.py:72: AttributeError
============================================================================================ warnings summary =============================================================================================
venv/lib64/python3.6/site-packages/_pytest/junitxml.py:436
  /home/z092789/code/treasury-ml/venv/lib64/python3.6/site-packages/_pytest/junitxml.py:436: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0.
  Add 'junit_family=legacy' to your pytest.ini file to silence this warning and make your suite compatible.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
---------------------------------------------------------------------- generated xml file: /home/z092789/code/treasury-ml/rspec.xml -----------------------------------------------------------------------
====================================================================================== 1 failed, 1 warning in 6.57s =======================================================================================
```

**Comment by tseader** [comment]
Another telling error when I replaced `df.collect()` with `df.show()`, namely:
```text
py4j.protocol.Py4JJavaError: An error occurred while calling o35.showString.
E                   : java.lang.IllegalStateException: SparkContext has been shutdown
```

# Change to test function
```python
def test_simple_pyspark_decorator():
    res = execute_solid(simple_solid, mode_def=ModeDefinition(
        resource_defs={'pyspark': pyspark_resource}))
    df: SparkDataFrame = res.output_value()
    assert isinstance(df, SparkDataFrame)
    print(df.show()) # <-- the only change
```

# Full output dump
```text
$ python -m pytest tests/int/test_dagster_spark.py -s -vvvv
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.6.8, pytest-5.3.1, py-1.8.0, pluggy-0.13.1 -- /home/z092789/code/treasury-ml/venv/bin/python
cachedir: .pytest_cache
rootdir: /home/z092789/code/treasury-ml, inifile: pytest.ini
plugins: cov-2.8.1, pylint-0.14.1, mockito-0.0.4
collected 1 item                                                                                                                                                                                          

tests/int/test_dagster_spark.py::test_simple_pyspark_decorator 19/12/11 12:21:22 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
2019-12-11 12:21:23 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7062f360-cb04-4d72-b1f6-f85cc6c296e5 - PIPELINE_START - Started execution of pipeline "ephemeral_simple_solid_solid_pipeline".
2019-12-11 12:21:23 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7062f360-cb04-4d72-b1f6-f85cc6c296e5 - ENGINE_EVENT - Executing steps in process (pid: 13315)
 event_specific_data = {"metadata_entries": [["pid", null, ["13315"]], ["step_keys", null, ["{'simple_solid.compute'}"]]]}
2019-12-11 12:21:23 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7062f360-cb04-4d72-b1f6-f85cc6c296e5 - STEP_START - Started execution of step "simple_solid.compute".
               solid = "simple_solid"
    solid_definition = "simple_solid"
            step_key = "simple_solid.compute"
2019-12-11 12:21:26 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7062f360-cb04-4d72-b1f6-f85cc6c296e5 - STEP_OUTPUT - Yielded output "result" of type "Any". (Type check passed).
 event_specific_data = {"intermediate_materialization": null, "step_output_handle": ["simple_solid.compute", "result"], "type_check_data": [true, "result", null, []]}
               solid = "simple_solid"
    solid_definition = "simple_solid"
            step_key = "simple_solid.compute"
2019-12-11 12:21:26 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7062f360-cb04-4d72-b1f6-f85cc6c296e5 - STEP_SUCCESS - Finished execution of step "simple_solid.compute" in 2.51s.
 event_specific_data = {"duration_ms": 2512.9907720256597}
               solid = "simple_solid"
    solid_definition = "simple_solid"
            step_key = "simple_solid.compute"
2019-12-11 12:21:26 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7062f360-cb04-4d72-b1f6-f85cc6c296e5 - ENGINE_EVENT - Finished steps in process (pid: 13315) in 2.52s
 event_specific_data = {"metadata_entries": [["pid", null, ["13315"]], ["step_keys", null, ["{'simple_solid.compute'}"]]]}
2019-12-11 12:21:26 - dagster - DEBUG - ephemeral_simple_solid_solid_pipeline - 7062f360-cb04-4d72-b1f6-f85cc6c296e5 - PIPELINE_SUCCESS - Finished execution of pipeline "ephemeral_simple_solid_solid_pipeline".
['col1', 'col2']
FAILED

================================================================================================ FAILURES =================================================================================================
______________________________________________________________________________________ test_simple_pyspark_decorator ______________________________________________________________________________________

    def test_simple_pyspark_decorator():
        res = execute_solid(simple_solid, mode_def=ModeDefinition(
            resource_defs={'pyspark': pyspark_resource}))
        df: SparkDataFrame = res.output_value()
        assert isinstance(df, SparkDataFrame)
        print(df.columns)
>       print(df.show())

tests/int/test_dagster_spark.py:21: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib64/python3.6/site-packages/pyspark/sql/dataframe.py:380: in show
    print(self._jdf.showString(n, 20, vertical))
venv/lib64/python3.6/site-packages/py4j/java_gateway.py:1257: in __call__
    answer, self.gateway_client, self.target_id, self.name)
venv/lib64/python3.6/site-packages/pyspark/sql/utils.py:63: in deco
    return f(*a, **kw)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

answer = 'xro58', gateway_client = <py4j.java_gateway.GatewayClient object at 0x7fa3de36c128>, target_id = 'o35', name = 'showString'

    def get_return_value(answer, gateway_client, target_id=None, name=None):
        """Converts an answer received from the Java gateway into a Python object.
    
        For example, string representation of integers are converted to Python
        integer, string representation of objects are converted to JavaObject
        instances, etc.
    
        :param answer: the string returned by the Java gateway
        :param gateway_client: the gateway client used to communicate with the Java
            Gateway. Only necessary if the answer is a reference (e.g., object,
            list, map)
        :param target_id: the name of the object from which the answer comes from
            (e.g., *object1* in `object1.hello()`). Optional.
        :param name: the name of the member from which the answer comes from
            (e.g., *hello* in `object1.hello()`). Optional.
        """
        if is_error(answer)[0]:
            if len(answer) > 1:
                type = answer[1]
                value = OUTPUT_CONVERTER[type](answer[2:], gateway_client)
                if answer[1] == REFERENCE_TYPE:
                    raise Py4JJavaError(
                        "An error occurred while calling {0}{1}{2}.\n".
>                       format(target_id, ".", name), value)
E                   py4j.protocol.Py4JJavaError: An error occurred while calling o35.showString.
E                   : java.lang.IllegalStateException: SparkContext has been shutdown
E                       at org.apache.spark.SparkContext.runJob(SparkContext.scala:2053)
E                       at org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)
E                       at org.apache.spark.SparkContext.runJob(SparkContext.scala:2101)
E                       at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:365)
E                       at org.apache.spark.sql.execution.CollectLimitExec.executeCollect(limit.scala:38)
E                       at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collectFromPlan(Dataset.scala:3389)
E                       at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2550)
E                       at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2550)
E                       at org.apache.spark.sql.Dataset$$anonfun$52.apply(Dataset.scala:3370)
E                       at org.apache.spark.sql.execution.SQLExecution$$anonfun$withNewExecutionId$1.apply(SQLExecution.scala:78)
E                       at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:125)
E                       at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:73)
E                       at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3369)
E                       at org.apache.spark.sql.Dataset.head(Dataset.scala:2550)
E                       at org.apache.spark.sql.Dataset.take(Dataset.scala:2764)
E                       at org.apache.spark.sql.Dataset.getRows(Dataset.scala:254)
E                       at org.apache.spark.sql.Dataset.showString(Dataset.scala:291)
E                       at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
E                       at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
E                       at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
E                       at java.lang.reflect.Method.invoke(Method.java:498)
E                       at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
E                       at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
E                       at py4j.Gateway.invoke(Gateway.java:282)
E                       at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
E                       at py4j.commands.CallCommand.execute(CallCommand.java:79)
E                       at py4j.GatewayConnection.run(GatewayConnection.java:238)
E                       at java.lang.Thread.run(Thread.java:748)

venv/lib64/python3.6/site-packages/py4j/protocol.py:328: Py4JJavaError
============================================================================================ warnings summary =============================================================================================
venv/lib64/python3.6/site-packages/_pytest/junitxml.py:436
  /home/z092789/code/treasury-ml/venv/lib64/python3.6/site-packages/_pytest/junitxml.py:436: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0.
  Add 'junit_family=legacy' to your pytest.ini file to silence this warning and make your suite compatible.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
---------------------------------------------------------------------- generated xml file: /home/z092789/code/treasury-ml/rspec.xml -----------------------------------------------------------------------
====================================================================================== 1 failed, 1 warning in 8.05s =======================================================================================
```

**Comment by natekupp** [comment]
Hey Tyler thanks for the issue report! I'll take a look at this today or tomorrow.

**Comment by natekupp** [comment]
This uncovered an issue we had with resources not being properly provisioned during solid output collection, which is fixed here:

https://dagster.phacility.com/D1643

This diff includes your test which is now passing.

Also, as noted in the diff, we currently call `.stop()` on the `SparkSession` by default during pipeline execution cleanup, which causes the `df.collect()` to throw an exception. I've added a config flag to disable this behavior, used in the test.

Finally, when fixing this I also noticed that we are currently hard-coding the pyspark resource name to `spark` in the pyspark DataFrame Dagster type. I filed #1983 for us to fix that in a follow-up.

This is now in master and should go out in a PyPI release soon. Thanks again for the great bug report!
